pkg/apis/sessions/session_state.go

```diff
@@ -66,64 +73,114 @@ func (s *SessionState) String() string {
-			}
-		}
-		if ss.IDToken != "" {
-			ss.IDToken, err = c.Encrypt(ss.IDToken)
-			if err != nil {
-				return "", err
-			}
+// EncodeSessionState returns an lz4 compression (optional) of a MessagePack encoded session
+// Encryption & Base64 encoding are delegated to downstream consumers of this method.
+func (s *SessionState) EncodeSessionState(compress bool) ([]byte, error) {
+	var (
+		ss  SessionState
+		err error
+
+		// LZ4 & MessagePack
+		packed []byte
+		reader *bytes.Reader
+		buf    *bytes.Buffer
+		zw     *lz4.Writer
+	)
+
+	ss = *s
+
+	// Embed SessionState, Decoded Tokens and Expires pointers into SessionStateCompressed
+	sse := &SessionStateEncoded{SessionState: &ss}
+	if !ss.CreatedAt.IsZero() {
+		sse.CreatedAt = &ss.CreatedAt
+	}
+	if !ss.ExpiresOn.IsZero() {
+		sse.ExpiresOn = &ss.ExpiresOn
+	}
+
+	//Marshal & Compress the SessionStateCompressed
+	packed, err = msgpack.Marshal(sse)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	if !compress {
+		return packed, nil
+	}
+
+	// The Compress:Decompress ratio is 1:Many. LZ4 gives fastest decompress speeds
+	buf = new(bytes.Buffer)
+	zw = lz4.NewWriter(nil)
+	zw.Header = lz4.Header{
+		BlockMaxSize:     65536,
+		CompressionLevel: 0,
+	}
+	zw.Reset(buf)
+
+	reader = bytes.NewReader(packed)
+	_, err = io.Copy(zw, reader)
+	if err != nil {
+		return []byte{}, err
+	}
+	_ = zw.Close()
+
+	return ioutil.ReadAll(buf)
+}
+
+// DecodeSessionState decodes a LZ4 compressed MessagePack into a Session State
+func DecodeSessionState(data []byte, compressed bool) (*SessionState, error) {
+	var (
+		sse SessionStateEncoded
+		ss  *SessionState
+		err error
+
+		// LZ4 & MessagePack
+		buf    *bytes.Buffer
+		reader *bytes.Reader
+		zr     *lz4.Reader
+		packed []byte
+	)
+
+	packed = data
+	if compressed {
+		reader = bytes.NewReader(data)
+		buf = new(bytes.Buffer)
+		zr = lz4.NewReader(nil)
+		zr.Reset(reader)
+		_, err = io.Copy(buf, zr)
+		if err != nil {
+			return nil, err
 		}
-		if ss.RefreshToken != "" {
-			ss.RefreshToken, err = c.Encrypt(ss.RefreshToken)
-			if err != nil {
-				return "", err
-			}
+
+		packed, err = ioutil.ReadAll(buf)
+		if err != nil {
+			return nil, err
 		}
 	}
-	// Embed SessionState and ExpiresOn pointer into SessionStateJSON
-	ssj := &SessionStateJSON{SessionState: &ss}
-	if !ss.CreatedAt.IsZero() {
-		ssj.CreatedAt = &ss.CreatedAt
+
+	err = msgpack.Unmarshal(packed, &sse)
+	if err != nil {
+		return nil, err
 	}
-	if !ss.ExpiresOn.IsZero() {
-		ssj.ExpiresOn = &ss.ExpiresOn
+	if sse.SessionState == nil {
+		return nil, fmt.Errorf("failed to decode the session state")
 	}
-	b, err := json.Marshal(ssj)
-	return string(b), err
+
+	ss = sse.SessionState
+	if sse.CreatedAt != nil {
+		ss.CreatedAt = *sse.CreatedAt
+	}
+	if sse.ExpiresOn != nil {
+		ss.ExpiresOn = *sse.ExpiresOn
+	}
+
+	// Holdover behavior from Legacy decode
+	// NOTE: this makes decode NOT a 1:1 reversal of Encode
+	// TODO: Is this the best place for this logic?
+	if ss.User == "" {
```

Comment:
What's the situtation with this?  I don't like how it makes `data != Decode(Encode(data))` in some cases.

Is this a legacy artifact?  If not, does it make sense to assign the user to the email if the user is missing outside of the Decode function so it can remain the inverse of Encode?